### ISSUE TITLE
Patching failing with Gateway Timeout

### ISSUE THREAD

**Issue by sajuptpm** [nolar/kopf#738]
* kopf: 1.29
* curl to https://10.0.0.1:443/apis/apple.io/v1/namespaces/default/apps/abc-1.0.0 is working from kopf pod
* kubectl client also working on that cluster, that means k8s api server is up and running.

===================== log =====================

[2021-04-16 06:06:08,466] kopf.objects[DEBUG   ] [default/abc-1.0.0] **Patching with:** {'status':  ......

2021-04-16 06:06:42,450] kopf.objects         [ERROR   ] [default/abc-1.0.0] Throttling for 1 seconds due to an unexpected error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/kopf/clients/errors.py", line 133, in check_response
    response.raise_for_status()
  File "/usr/local/lib/python3.7/site-packages/aiohttp/client_reqrep.py", line 1005, in raise_for_status
    headers=self.headers,
**aiohttp.client_exceptions.ClientResponseError: 504, message='Gateway Timeout', url=URL('https://10.0.0.1:443/apis/apple.io/v1/namespaces/default/apps/abc-1.0.0')**

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/kopf/reactor/effects.py", line 200, in throttled
    yield should_run
  File "/usr/local/lib/python3.7/site-packages/kopf/reactor/processing.py", line 102, in process_resource_event
    replenished=replenished,
  File "/usr/local/lib/python3.7/site-packages/kopf/reactor/effects.py", line 62, in apply
    await patch_and_check(resource=resource, patch=patch, body=body, logger=logger)
  File "/usr/local/lib/python3.7/site-packages/kopf/reactor/effects.py", line 124, in patch_and_check
    patch=patch,
  File "/usr/local/lib/python3.7/site-packages/kopf/clients/auth.py", line 44, in wrapper
    return await fn(*args, **kwargs, context=context)
  File "/usr/local/lib/python3.7/site-packages/kopf/clients/patching.py", line 53, in patch_obj
    patched_body = await errors.parse_response(response)
  File "/usr/local/lib/python3.7/site-packages/kopf/clients/errors.py", line 144, in parse_response
    await check_response(response)
  File "/usr/local/lib/python3.7/site-packages/kopf/clients/errors.py", line 135, in check_response
    raise cls(payload, status=response.status) from e
kopf.clients.errors.APIError: ('**Timeout: request did not complete within requested timeout 34s**', {'kind': 'Status', 'apiVersion': 'v1', 'metadata': {}, 'status': 'Failure', 'message': 'Timeout: request did not complete within requested timeout 34s', 'reason': 'Timeout', 'details': {}, 'code': 504})

**Comment by nolar** [comment]
That looks like an error inside the Kubernetes API servers setup, Kopf can do nothing to fix it. I.e., it is not even Kopf's client<->server timeout, it is the remote timeout, probably between the load balancers and the actual API servers (all are parts of Kubernetes).

is this error happening in the pod? The same where `curl`/`kubectl` is called from? Does it happen when `curl` is tried several times? With the same method and similar payload (the patch body)? Is there anything suspicious in the logs of the API servers / load-balancers?

**Comment by Bhaumik10** [comment]
kopf: version 1.2 & above
Kubernetes Version: 1.16
Cluster: AWS EKS (Managed service from EKS)
=========================== Issue Description =========================
`Issue`: When try to delete the application object, `kopf` framework internally had an error.You can see Error 500 in attached log snapshot. Due to that state of the application object cached internally and not cleared out though framework didn't executed the intended deletion. This leads to the behavior where we try to do force delete from command line though object won't get deleted.

I have attached error snapshots below.

![image](https://user-images.githubusercontent.com/9876769/117711399-0e637600-b188-11eb-8e6e-8a09976e80ee.png)



Suggestion: What we are thinking is, if framework can identify/detect such failure and add retry for may be 3 times. If succeed then return response else raise exception. So, as user we handle our application appropriately. 

@sajuptpm do you see such behavior working with `EKS + kopf`? or you have on-premise/self managed kubernetes cluster and seeing such behavior?

**Comment by nolar** [comment]
On a second thought, this might be a good idea. 

I’ll add a configurable number of retries before giving up (to settings, default 1) — for patching & watching & other api operations.

The only question now is which kinds of errors should be covered besides TCP connectivity. All 500s?

**Comment by Bhaumik10** [comment]
Hi @nolar,
Thanks for responding on this issue. I would suggest we should retry on all `500 + lower level protocols on framework side`. Also, try to use exponential backoff when we do retry. So, if something is transient then wait between retry will help.
we do use following package `https://pypi.org/project/retry/` to achieve retry with effective tries between subsequent requests.
```
import retry import retry
from datetime import datetime

@retry(exceptions=(Exception), tries=3, delay=1, backoff=4)
def test_retry():
	print(f"current timestamp: {datetime.now()}")
	raise Exception("Testing retry Package")

Note: You can specify specific exception to handle. I have tried retry with any exception here.
```

**Comment by donbowman** [comment]
This also occurrs when using istio sidecars. The ability to get to the kubernetes api can be missing on first startup (until the sidecar sorts itself out)

`aiohttp.client_exceptions.ClientResponseError: 503, message='Service Unavailable', url=URL('https://10.27.240.1:443/apis/metrics.k8s.io/v1beta1')`

comes out, then we exit, then we come up, same error, repeat.

**Comment by nolar** [comment]
@donbowman In your case, the URL does not look like it is used for PATCHing. Maybe it is closer to #740 (solved in #741, [released as](https://github.com/nolar/kopf/releases) 1.32rc1/1.32rc2)?

**Comment by donbowman** [comment]
is it not that generically that a 50x can occur on the kubernetes api and we need a backoff/retry?

in my case it happens because the istio sidecar is not yet aval, causing a 503.
but, it could happen a bit later since e.g. if the network connection to the api server drops, the sidecar would generate a 503 (for any operation, patch included)

**Comment by nolar** [comment]
I was working on a different performance/scalability issue, but that one has exploded in complexity and possible branches of a decision tree. I guess I will take a pause there and do these API retries instead.

**Comment by Bhaumik10** [comment]
hi @nolar 
Just want to follow up on this issue. Do we have fix out yet or still it's in progress?

**Comment by Bhaumik10** [comment]
@nolar we are continuously getting into situation where though operator object/pod is deleted still it tries to delete it and ending up with 409s.`ApiException: (409)`

<img width="1633" alt="error_message" src="https://user-images.githubusercontent.com/9876769/122457664-bcbac200-cf63-11eb-8844-a17367d198e3.png">

**Comment by nolar** [comment]
The timeouts & retries configuration is released as `kopf>=1.33.0`. Documentation: https://kopf.readthedocs.io/en/stable/configuration/#retrying-of-api-errors

This covers all internal API operations of the framework: watching, patching, peering touching, cluster scanning, etc. But not of the handlers — they should care about their own API errors or rely on Kopf's generic handler retrying.

I close this issue as solved. Feel free to re-open if something is left unsolved regarding the timeouts/server-side/networking errors.

@Bhaumik10 For the "409 Conflict" while deleting — this seems like a different error. 4xx errors mean that "the client is wrong". Generally, such errors should not be retried. Can you please create a separate issue with an example, logs, and an exact scenario to reproduce?
